stream-chat-android-ui-common/src/main/kotlin/com/getstream/sdk/chat/utils/typing/DefaultTypingUpdatesBuffer.kt

```diff
@@ -0,0 +1,143 @@
+ */
+public class DefaultTypingUpdatesBuffer(
+    private val onTypingStarted: () -> Unit,
+    private val onTypingStopped: () -> Unit,
+) : TypingUpdatesBuffer {
+
+    /**
+     * The coroutine scope used for running the timer and
+     * sending updates.
+     */
+    private val coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
+
+    /**
+     * Holds the currently running job.
+     */
+    private var isTypingTimerJob: Job? = null
+
+    /**
+     * Used to intelligently call [onTypingStopped] and [onTypingStopped].
+     */
+    private var sendUpdatesJob: Job? = null
+
+    /**
+     * If the user is currently typing or not.
+     *
+     * Setting this value automatically triggers handling
+     * typing events.
+     */
+    private var isTyping: Boolean = false
+        set(value) {
+            field = value
+            if (isTyping) {
+                handleTypingEvent()
+            }
+        }
+
+    /**
+     * Flips [isTyping] to false after a set amount of time
+     * dictated by [DEFAULT_BUFFER_TYPING_UPDATES_INTERVAL].
+     */
+    private suspend fun startTypingTimer() {
+        delay(DEFAULT_BUFFER_TYPING_UPDATES_INTERVAL)
+        isTyping = false
+    }
+
+    /**
+     * Sets the value of [isTyping] only if there is
+     * a change in state in order to not create unnecessary events.
+     *
+     * Also resets the job used to time periods of typing activity.
+     */
+    override fun onKeystroke() {
+        if (!isTyping) {
+            isTyping = true
+        }
+        isTypingTimerJob?.cancel()
+        isTypingTimerJob = coroutineScope.launch { startTypingTimer() }
+    }
+
+    /**
+     * Sets [isTyping] to false and cancels [coroutineScope].
+     *
+     * Should be called before an instance of this class is removed.
+     */
+    override fun clear() {
+        if (isTyping) {
+            isTyping = false
+        }
+        coroutineScope.cancel()
+    }
+
+    /**
+     * Calls [onTypingStarted] every [DEFAULT_SEND_TYPING_UPDATES_INTERVAL] while
+     * the user is still typing and calls [onTypingStopped] afterwards.
+     */
+    private fun handleTypingEvent() {
+        sendUpdatesJob?.cancel()
+
+        sendUpdatesJob = coroutineScope.launch {
+
+            do {
+                onTypingStarted()
+                delay(DEFAULT_SEND_TYPING_UPDATES_INTERVAL)
+            } while (isTyping)
+
+            onTypingStopped()
+        }
+    }
+
+    private companion object {
+        /**
+         * Dictates the frequency start typing updates are sent
+         * at while the user is typing.
+         */
+        private const val DEFAULT_SEND_TYPING_UPDATES_INTERVAL: Long = 3000L
+
+        /**
+         * Dictates how long the user has to stop typing for
+         * before he is considered to be inactive (stopped typing).
+         */
+        private const val DEFAULT_BUFFER_TYPING_UPDATES_INTERVAL: Long = 1000L
```

Comment:
This is set to `DEFAULT_SEND_TYPING_UPDATES_INTERVAL/ n` so that we can minimize the chances of overrun and doing another loop unnecessarily.